### PR TITLE
feat!: Iterator `Key()` and `Value()` no longer return a copy

### LINKED ISSUE #156: Make a second DB "IterKey" API that returns a key to the caller that can potentially mutate
Currently the Iterator.Key() API makes a copy of the key it gets from the database. This is because the database's iterator returns something it will mutate on the subsequent .Next() call for heap efficiency. This extra copy causes very large heap allocation (and time overheads) to query serving nodes, and a 1% time overhead to the entire state machine time for Osmosis.

On a heap allocation profile of a query serving Osmosis RPC node over an hour, it has 450 gigabytes allocated from this API. On spot-check, none of the big ones need this copying behavior. (160GB removed from a tendermint update, but the remaining 290GB are still from this API)

image

In the state machine, we see 1% of state machine execution time is blocked on copying this key, again in situations where I don't think we need any of this either. 
image

-------------

Proposal: Add a new method `KeyMut()` to the interface for Iterator. The caller should not mutate this key, and the expectation is that the key may get mutated on the next `.Next()` call. 

I'm not stoked about the naming of this method, so happy for better ideas

### PR DESCRIPTION
**Pull request by alesforz**
### Context
Closes #156.

### Changes

The:
- `Iterator.Key()` API implementations of levelDB, badgerDB, and pebbleDB no longer return a copy of the key.
- `Iterator.Value()` API implementations of levelDB and pebbleDB no longer return a copy of the value.

Rather, as the updated docs reflect, the caller is responsible for creating a copy of the slice if they want to modify it.

See my further comments about [rocksdb](https://github.com/cometbft/cometbft-db/pull/168#issuecomment-2222932698) and [badgerdb](https://github.com/cometbft/cometbft-db/pull/168#issuecomment-2223080924) below.

### CODE REVIEW

**Comment by alesforz** [comment]
Unfortunately, I don't think we can avoid the copy in rocksdb `Key()` implementation.
Under the hood, the library we use allocates and returns a [C array](https://github.com/linxGnu/grocksdb/blob/v1.8.14/slice.go#L9), which is subject to C's memory resource policy.
But Go's memory management can't keep track of any memory allocated in the C library.
Therefore, to ensure there are no memory leaks call `Free()`, as the current implementation [does](https://github.com/cometbft/cometbft-db/blob/7ed49f2a912b5e366e3ac9847c25170b6a97bb74/rocksdb_iterator.go#L139).

In short, my understanding is that we need the copy of the key in order to free its memory allocated in the C library, thus avoiding possible memory leaks.

**Comment by alesforz** [comment]
We can't avoid the copy in the `Value()` API of badgerDB either.
Namely, badgerDB exposes 2 APIs to retrieve a value:
```go
func (item *Item) Value(fn func(val []byte) error) error
func (item *Item) ValueCopy(dst []byte) ([]byte, error)
```

Note:
- `Item.Value` does not return the value. Instead, it wants a closure that works on it directly inside the library's boundaries. The `Value` method will call `fn` passing the value we want to retrieve as an argument `[]byte`. To extract it we must do something like:
  ```go
  var valueCopy []byte
  err = item.Value(func(val []byte) error {
      // Make a copy of the value.
      valueCopy = append([]byte{}, val...)
      return nil
  })
  ```
  That is, we have to manually copy the value.
- `Item.ValueCopy` creates a copy for us and returns it.

In both cases, the library returns as a copy of the original value; it offers no possibility of accessing the original value directly.

**Comment by alesforz** [comment]
Given the discussion above, addressing #168 makes the library behave inconsistently depending on the underlying database used. Specifically, some databases require callers to make copies of key and value, while others do not. Consequently, users cannot switch between databases without first checking the behavior of their `Key()` and `Value()` APIs and adapting their code accordingly.

File: badger_db.go [hunk 1]
```diff
@@ -279,19 +279,22 @@ func (i *badgerDBIterator) Valid() bool {
...
 	if !i.Valid() {
 		panic("iterator is invalid")
 	}
-	// Note that we don't use KeyCopy, so this is only valid until the next
-	// call to Next.
-	return i.iter.Item().KeyCopy(nil)
+	return i.iter.Item().Key()
 }
 
+// Value implements Iterator.
```
Comment (melekes) [inline badger_db.go:292]: we should indicate that Value is a copy

File: badger_db.go [hunk 2]
```diff
@@ -279,19 +279,22 @@ func (i *badgerDBIterator) Valid() bool {
 	return true
 }
 
+// Key implements Iterator.
```
Comment (melekes) [inline badger_db.go:282]:
```suggestion
```

I'd remove these comments.

**Review by melekes — COMMENTED** [review]
looks good so far 👍

**Comment by melekes** [comment]
> Note: this is a breaking change.

That's debatable because, in the `Iterator` documentation we say `read-only []byte`, although we don't mention anything about the lifetime of the pointer.

**Comment by alesforz** [comment]
> > Note: this is a breaking change.
> 
> That's debatable because, in the `Iterator` documentation we say `read-only []byte`, although we don't mention anything about the lifetime of the pointer.

I see your point. 
The `Iterator` interface docs states:
```
// As with DB, keys and values should be considered read-only, and must be copied before they are
// modified.
```
therefore, users of the library know that they must create a copy, thus these changes shouldn't break anything.

File: badger_db.go (hunk 2 above)
Comment (alesforz) [reply badger_db.go:282]: 🤔 I would leave them, because I think they give readers some additional context about the functions. 
Granted, in the context of this repo it's clear what the functions refer to. 
Still, I would argue it doesn't hurt to leave extra-info in.

**Review by melekes — APPROVED** [review]
👍

**Comment by melekes** [comment]
Could you add a changelog entry? btw we're using https://github.com/informalsystems/unclog
